File: examples/platforms/kotlin/src/test/kotlin/com/boltffi/demo/CallbackErrorsTest.kt
package com.boltffi.demo

import kotlinx.coroutines.runBlocking
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith

class CallbackErrorsTest {
    private fun checkMode(mode: Int) {
        when (mode) {
            1 -> throw MathError.NegativeInput
            2 -> error("unchecked callback failure 東京\u0000🦀")
        }
    }

    private val worker = object : FallibleWorker {
        override fun run(mode: Int) = checkMode(mode)
        override fun value(mode: Int): Int {
            checkMode(mode)
            return 42
        }
    }

    private val asyncWorker = object : AsyncFallibleWorker {
        override suspend fun run(mode: Int) = checkMode(mode)
        override suspend fun value(mode: Int): Int {
            checkMode(mode)
            return 42
        }
    }

    @Test
    fun callbackErrorsReachRust() = runBlocking {
        demoCase("case:callbacks.errors.unit.should_report_success")
        assertEquals(Unit, invokeUnitWorker(worker, 0))
        demoCase("case:callbacks.errors.unit.should_report_declared_error")
        assertEquals(MathError.NegativeInput, assertFailsWith<MathError> { invokeUnitWorker(worker, 1) })
        demoCase("case:callbacks.errors.unit.should_report_unexpected_error")
        assertEquals(MathError.Overflow, assertFailsWith<MathError> { invokeUnitWorker(worker, 2) })
        demoCase("case:callbacks.errors.value.should_report_success")
        assertEquals(42, invokeValueWorker(worker, 0))
        demoCase("case:callbacks.errors.value.should_report_declared_error")
        assertEquals(MathError.NegativeInput, assertFailsWith<MathError> { invokeValueWorker(worker, 1) })
        demoCase("case:callbacks.errors.value.should_report_unexpected_error")
        assertEquals(MathError.Overflow, assertFailsWith<MathError> { invokeValueWorker(worker, 2) })
        demoCase("case:callbacks.errors.async_unit.should_report_success")
        assertEquals(Unit, invokeAsyncUnitWorker(asyncWorker, 0))
        demoCase("case:callbacks.errors.async_unit.should_report_declared_error")
        assertEquals(MathError.NegativeInput, assertFailsWith<MathError> { invokeAsyncUnitWorker(asyncWorker, 1) })
        demoCase("case:callbacks.errors.async_unit.should_report_unexpected_error")
        assertEquals(MathError.Overflow, assertFailsWith<MathError> { invokeAsyncUnitWorker(asyncWorker, 2) })
        demoCase("case:callbacks.errors.async_value.should_report_success")
        assertEquals(42, invokeAsyncValueWorker(asyncWorker, 0))
        demoCase("case:callbacks.errors.async_value.should_report_declared_error")
        assertEquals(MathError.NegativeInput, assertFailsWith<MathError> { invokeAsyncValueWorker(asyncWorker, 1) })
        demoCase("case:callbacks.errors.async_value.should_report_unexpected_error")
        assertEquals(MathError.Overflow, assertFailsWith<MathError> { invokeAsyncValueWorker(asyncWorker, 2) })
        assertEquals(MathError.Overflow, assertFailsWith<MathError> {
            applyResultClosure(ClosureI32ToResultI32ErrMathError { error("unexpected closure error") }, 0)
        })
    }
}

File: examples/platforms/dart/test/callback_errors_test.dart
import 'package:demo/demo.dart';
import 'package:test/test.dart';

final class Worker implements FallibleWorker {
  @override
  void run(int mode) {
    if (mode == 1) throw MathError.negativeInput;
    if (mode == 2) throw StateError('unchecked callback failure 東京🦀');
  }

  @override
  int value(int mode) {
    run(mode);
    return 42;
  }
}

final class AsyncWorker implements AsyncFallibleWorker {
  final worker = Worker();

  @override
  Future<void> run(int mode) async => worker.run(mode);

  @override
  Future<int> value(int mode) async => worker.value(mode);
}

void main() {
  tearDownAll(shutdownBoltffi);
  test('callback errors reach Rust', () async {
    final worker = Worker();
    final asyncWorker = AsyncWorker();
    expect(
      () => invokeUnitWorker(worker, 0),
      returnsNormally,
      reason: 'case:callbacks.errors.unit.should_report_success',
    );
    expect(
      () => invokeUnitWorker(worker, 1),
      throwsA(MathError.negativeInput),
      reason: 'case:callbacks.errors.unit.should_report_declared_error',
    );
    expect(
      () => invokeUnitWorker(worker, 2),
      throwsA(MathError.overflow),
      reason: 'case:callbacks.errors.unit.should_report_unexpected_error',
    );
    expect(
      invokeValueWorker(worker, 0),
      42,
      reason: 'case:callbacks.errors.value.should_report_success',
    );
    expect(
      () => invokeValueWorker(worker, 1),
      throwsA(MathError.negativeInput),
      reason: 'case:callbacks.errors.value.should_report_declared_error',
    );
    expect(
      () => invokeValueWorker(worker, 2),
      throwsA(MathError.overflow),
      reason: 'case:callbacks.errors.value.should_report_unexpected_error',
    );
    await expectLater(
      invokeAsyncUnitWorker(asyncWorker, 0),
      completes,
      reason: 'case:callbacks.errors.async_unit.should_report_success',
    );
    await expectLater(
      invokeAsyncUnitWorker(asyncWorker, 1),
      throwsA(MathError.negativeInput),
      reason: 'case:callbacks.errors.async_unit.should_report_declared_error',
    );
    await expectLater(
      invokeAsyncUnitWorker(asyncWorker, 2),
      throwsA(MathError.overflow),
      reason: 'case:callbacks.errors.async_unit.should_report_unexpected_error',
    );
    expect(
      await invokeAsyncValueWorker(asyncWorker, 0),
      42,
      reason: 'case:callbacks.errors.async_value.should_report_success',
    );
    await expectLater(
      invokeAsyncValueWorker(asyncWorker, 1),
      throwsA(MathError.negativeInput),
      reason: 'case:callbacks.errors.async_value.should_report_declared_error',
    );
    await expectLater(
      invokeAsyncValueWorker(asyncWorker, 2),
      throwsA(MathError.overflow),
      reason:
          'case:callbacks.errors.async_value.should_report_unexpected_error',
    );
    expect(
      () => applyResultClosure(
        (value) => throw StateError('unexpected closure error'),
        0,
      ),
      throwsA(MathError.overflow),
    );
  });
}
